quiz/quiz.go

```diff
@@ -0,0 +1,168 @@
+func New(filename string) (*Quiz, error) {
+	fmt.Println("\nGetting quiz using file:\t", filename)
+
+	// open the file and save Question to a Quiz struct
+	csvFile, err := os.Open(filename)
+
+	if err != nil {
+		fmt.Println("Error opening CSV file:\t", err)
+		return nil, err
+	}
+
+	defer func() {
+		if csvErr := csvFile.Close(); csvErr != nil {
+			fmt.Println(csvErr)
+		}
+	}()
+
+	reader := csv.NewReader(bufio.NewReader(csvFile))
+
+	var questions []Question
+	// loop until end of file or error is reached
+	for {
+		record, ioErr := reader.Read()
+		if ioErr == io.EOF {
+			break
+		}
+		if ioErr != nil {
+			fmt.Println("Error reading from file:\t", ioErr)
+			return nil, ioErr
+		}
+
+		questions = append(questions, Question{
+			QuestionText: record[0],
+			Answer:       record[1],
+		})
+	}
+
+	return &Quiz{
+		Questions:        questions,
+		Filename:         filename,
+		TotalQuestions:   len(questions),
+		CorrectQuestions: 0,
+	}, err
+}
+
+// Start begins the user-input portion of the quiz. Each Question is presented to the user and they enter answers via
+// stdin with "\n" being used to specify the end of an answer.
+func (q *Quiz) Start(maxQuizTime time.Duration, random bool) error {
+	fmt.Printf("\nYou will have %v seconds to answer %v questions.\nPress [Enter] to Start the quiz.\n",
+		maxQuizTime, q.TotalQuestions)
+
+	reader := bufio.NewReader(os.Stdin)
+	_, err := reader.ReadString('\n')
+	if err != nil {
+		return err
+	}
+
+	completed := make(chan bool)
+
+	time.AfterFunc(maxQuizTime, func() {
+		fmt.Println("\nTimer has run out!")
+		completed <- false
+	})
+
+	if random {
+		go proctorRandomQuiz(q, reader, completed)
+	} else {
+		go proctorQuiz(q, reader, completed)
+	}
+
+	if <-completed {
+		return nil
+	} else {
+		errorText := fmt.Sprintf("The quiz timer (%v seconds) ran out before the user completed the quiz",
+			maxQuizTime)
+		return errors.New(errorText)
+	}
+}
+
+// proctorQuiz performs the quiz, asking the user questions and gathering responses. When it finishes, it sends
+// a "success" message down the channel to be returned to the calling function
+func proctorQuiz(q *Quiz, reader *bufio.Reader, c chan bool) {
+	for index, value := range q.Questions {
+		processQuestion(reader, &value, q, index)
+	}
+	c <- false
+}
+
+func proctorRandomQuiz(q *Quiz, reader *bufio.Reader, c chan bool) {
+	questionNumber := 1
+	for len(q.Questions) > 0 {
+		// gather a random index to get a random question
+		randomIndex := rand.Intn(len(q.Questions))
+
+		// store the random question
+		randomQuestion := q.Questions[randomIndex]
+
+		// remove the random question from the Questions slice
+		frontSlice := q.Questions[0:randomIndex]
+		backSlice := q.Questions[randomIndex+1 : len(q.Questions)]
+		q.Questions = append(frontSlice, backSlice...)
+
+		processQuestion(reader, &randomQuestion, q, questionNumber)
+		questionNumber++
+	}
+	c <- true
```

Comment:
you push a true on this channel, but what does it mean? You only check for something above, not it's value.